src/sync/sync_permission.hpp

```diff
@@ -0,0 +1,129 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_SYNC_PERMISSION_HPP
+#define REALM_OS_SYNC_PERMISSION_HPP
+
+#include "results.hpp"
+
+#include <string>
+
+namespace realm {
+
+class Permissions;
+class SyncUser;
+class Object;
+
+// Permission object used to represent a user permission
+struct Permission {
+    // The path of the Realm to which this permission pertains.
+    std::string path;
+
+    // A permission encapsulates a single access level.
+    // Each level includes all the capabilities of the level
+    // above it (for example, 'write' implies 'read').
+    enum class AccessLevel {
+        None,
+        Read,
+        Write,
+        Admin,
+    };
+    AccessLevel access;
+
+    // Condition is a userId or a KeyValue pair
+    // Other conditions may be supported in the future
+    struct Condition {
+        enum class Type { UserId, KeyValue };
+        Type type;
+
+        union {
+            std::string user_id;
+            std::pair<std::string, std::string> key_value;
+        };
+
+        Condition(std::string id) : type(Type::UserId), user_id(id) {}
+
+        Condition& operator=(const Condition&);
+        Condition(const Condition &c);
+        ~Condition();
+    };
+    Condition condition;
+
+    Permission(const Permission& p);
+    Permission(std::string path, AccessLevel access, Condition condition);
+    Permission& operator=(const Permission& p);
+};
+
+class PermissionResults {
+public:
+    // The number of permissions represented by this PermissionResults.
+    size_t size();
+
+    // Get the permission for the given index
+    // Throws OutOfBoundsIndexException if index >= size()
+    const Permission get(size_t index);
+
+    // Create an async query from this Results
+    // The query will be run on a background thread and delivered to the callback,
+    // and then rerun after each commit (if needed) and redelivered if it changed
+    NotificationToken async(std::function<void (std::exception_ptr)> target) { return m_results->async(target); }
+
+    // Create a new PermissionResults by further filtering or sorting this PermissionResults
+    PermissionResults filter(Query&& q) const;
+
+    // Create with a results - should be private
+    PermissionResults(std::unique_ptr<Results> results) : m_results(std::move(results)) {}
+
+private:
+    std::unique_ptr<Results> m_results;
```

Comment:
> How would you suggest move-constructing a Results?
```c++
new Results(std::move(some_results));
```
[The JS binding already does this when dealing with object store types](https://github.com/realm/realm-js/blob/92c10c469c5d92c4044d953e9167afa477dc687c/src/js_results.hpp#L98-L101) for the reason you mention.

> I was also having issues at one point with the move constructor for some object store types where the m_realm member or notifications weren't getting copied which is why I was using unique_ptrs in some spots.

Notifications will seem to disappear if you _copy_ a `Results` (or a `List` or `Object`) then discard the original as they're tied to a specific instance. If you move the `Results` they'll come along for the ride, though.